File: srcp/day-4.py
import time
from srcp.utils.input import get_data

data = get_data(2024, 4).strip()
start = time.time()

result_1 = result_2 = 0
grid: list[list[str]] = [list(line) for line in data.split("\n")]
rows, cols = len(lines := data.split()), len(lines)

list_x: list[tuple[int, int]] = []
list_center: list[tuple[int, int]] = []
list_dir: tuple[int, ...] = (-1, 0, 1)
word: str = "XMAS"

for r in range(rows):
    for c in range(cols):
        list_x += [(r, c)] if grid[r][c] == "X" else []
        list_center += [(r, c)] if grid[r][c] == "A" else []

for rx, cx in list_x:
    for di in list_dir:
        for dj in list_dir:
            result_1 += all(
                0 <= (ty := rx + k * di) < rows and
                0 <= (tx := cx + k * dj) < cols and
                grid[ty][tx] == c
                for k, c in enumerate(word)
            )

print(f"Result 1: {result_1} in {time.time() - start:.2f} s")
time_int = time.time()

for ca, ra in list_center:
    if (
            (xl := ra - 1) >= 0 and # L pour left
            (xr := ra + 1) < cols and # R pour right
            (yt := ca - 1) >= 0 and # T pour top
            (yb := ca + 1) < rows # B pour bottom
    ):
        diag_1 = {grid[yt][xl], grid[yb][xr]}
        diag_2 = {grid[yt][xr], grid[yb][xl]}

        result_2 += (diag_1 == diag_2 == set("MS"))

print(f"Result 2: {result_2} in {time.time() - time_int:.2f} s")

File: srcr/src/days/day4.rs
use std::collections::HashSet;

pub fn solve_day_4(input: &str) {
    let mut result_1 = 0;
    let mut result_2 = 0;

    let lines: Vec<&str> = input.trim().lines().collect();
    let grid: Vec<Vec<char>> = lines.iter().map(|line| line.chars().collect()).collect();

    let word = "XMAS";
    let mut list_x: Vec<(usize, usize)> = Vec::new();
    let mut list_center: Vec<(usize, usize)> = Vec::new();
    let list_dir: [isize; 3] = [-1, 0, 1];

    for r in 0..grid.len() {
        for c in 0..grid[0].len() {
            if grid[r][c] == 'X' {
                list_x.push((r, c));
            }
            if grid[r][c] == 'A' {
                list_center.push((r, c));
            }
        }
    }

    for (rx, cx) in list_x {
        for &di in &list_dir {
            for &dj in &list_dir {
                if di == 0 && dj == 0 {
                    continue;
                }
                if word.chars().enumerate().all(|(k, c)| {
                    let ty = rx as isize + k as isize * di;
                    let tx = cx as isize + k as isize * dj;
                    ty >= 0 && ty < grid.len() as isize &&
                        tx >= 0 && tx < grid[0].len() as isize &&
                        grid[ty as usize][tx as usize] == c
                }) {
                    result_1 += 1;
                }
            }
        }
    }

    for (ca, ra) in list_center {
        let xl = ra as isize - 1;
        let xr = ra as isize + 1;
        let yt = ca as isize - 1;
        let yb = ca as isize + 1;

        if xl >= 0 && xr < grid[0].len() as isize && yt >= 0 && yb < grid.len() as isize {
            let check = ['M', 'S'];
            let mut diag_1 = [
                grid[yt as usize][xl as usize],
                grid[yb as usize][xr as usize],
            ];
            let mut diag_2 = [
                grid[yt as usize][xr as usize],
                grid[yb as usize][xl as usize],
            ];
            diag_1.sort();
            diag_2.sort();

            if diag_1 == check && diag_2 == check {
                result_2 += 1;
            }
        }
    }

    println!("Result 1: {}", result_1);
    println!("Result 2: {}", result_2);
}